src/result.rs

```diff
@@ -1,43 +1,72 @@
+use core::fmt;
+
+pub type TockResult<T> = Result<T, TockError>;
+
+#[derive(Copy, Clone)]
+pub enum TockError {
```

Comment:
TockError is *huge*. SubscribeError alone is 3 words, plus there's a descriminant. That makes TockResult (for any non-empty type) at least 5 words.

Note that the ARM C ABI allows functions to return 2 word values. I don't know about RISC-V, but this is impractical for ARM.

Please shrink the size of these structs. Something more in line with the kernel's ReturnCode would be probably be a better fit here.

Reply:
It's even more. `CommandError` has 5 word-size fields. We could remove the `arg1` and `arg2` parameters, s.t. we end up with 3 fields. But without the driver number, minor number and return code, how should any developer find out what went wrong if it goes wrong?

Besides that, I would expect that the fields won't get compiled into the code unless you really care for them.

Why is 3 words plus a discriminant 5 words? Shouldn't a discriminant of 1 byte (maybe 1 word including alignment) be sufficient?

Reply:
> But without the driver number, minor number and return code, how should any developer find out what went wrong if it goes wrong?

This needs to be handled either through the type system (having different error types for different failures) or by having the code reporting the failure indicate where the failure came from.

> Besides that, I would expect that the fields won't get compiled into the code unless you really care for them.

`rustc` doesn't do this sort of layout optimization, and I wouldn't expect it to.

> Why is 3 words plus a discriminant 5 words? Shouldn't a discriminant of 1 byte (maybe 1 word including alignment) be sufficient?

`SubscribeError` is 3 words, aligned to the word (on most or all platforms Tock is interested in). `TockError` needs to add a descriminant to that, and that discriminant needs to be 1 word for alignment. `TockResult`, with a non-empty `T`, adds its own discriminant which also needs to be an entire word. Hence `3 + 1 + 1` words.

Reply:
I have two remarks on that discussion:
Firstly, the technical one. It would be nice to have all error information (except for possibly the return value from the kernel which cannot be know _a_ _priori_) covered by algebraic data types. I would expect usability to better then (because no `usize`, e.g. driver_number, values would have to be interpreted but their semantics is clear). However, I fear that the design and architecture will suffer then: I imagine an error type generic in a "driver" parameter or in a god-module knowing all drivers. Moreover, I believe (from my experience) that optimized code is much harder to test and to modify than non-optimized code.

Secondly, a meta point: I have the feeling that we are doing micro-optimization here: I'm not strongly against this (as I would be in the enterprise world where I originate) - I know resources _are_ limited. However, for objections concerning binary, size, stack consumption, performance I have the following desires:

 * actual examples where binary size is affected significantly,
 * actual examples where a performance penalty is measurable.

Moreover, ideally, I would like to
 
 * have tools to measure the impacts of code changes (i.e. metrics agreed upon),
 * have them tested automatically to identify regression in these non-functional requirements on each PR.

Having said that, I come to the following conclusion: If from the design perspective the Resultification is okay (and an algebraic version of the types in no better), I would go for it, unless actual users of the library are getting trouble with it. Otherwise we could optimize (which of course we could also do later).

Reply:
We don't currently have any code that uses TockResult/TockValue, so we don't have a size impact measurement for this. I'm fine with merging this as-is as long as we can change it in the future if code bloat proves to be a problem.

I'm working on code bloat evaluation tooling but it's going somewhat slow; I'll probably have something workable in January.

Reply:
> We don't currently have any code that uses TockResult/TockValue, so we don't have a size 
> impact measurement for this. I'm fine with merging this as-is as long as we can change it in
> the future if code bloat proves to be a problem.

I believe we will do some iterations with the API, anyway. And I think it's a good idea to change it when it gets a problem.

> I'm working on code bloat evaluation tooling but it's going somewhat slow; I'll probably 
> have something workable in January.

That would be great to have!

Reply:
Sorry for my late response!

Before merging this PR, I would like to mention that I already investigated a statically typed `Result` model 1 1/2 years ago (#46). At first sight, this felt pretty idiomatic and straight-forward. In the end, it became too clumsy and bloaty. The problems were:

- For more complex functions the return types got complex as well, e.g. `ButtonsError = GetDriverFailed(TockError) | NotSupported | SubscriptionFailed | SubscriptionFailed(TockError)`.
- Besides that, each error type revaled the internal implementation of a function. As a consequence, you needed to update your error type signature whenever the implementation was changed.
- You needed to `impl From` for each error type, s.t. it can be used in combination with the `?` operator.
- The typed results did not provide the information you needed. Usually, in a production scenario you will not try to dispatch every possible error since there are only very limited possibilities of handling them, if any. In a development situation, however, you might run into any possible error, e.g. due to a programming mistake. Given that you cannot tell which error will pop up before running the program, having a static error model would be very inconvenient.